[TEST-FAILURE] SlowOperationDetectorBasicTest.testSlowRecursiveOperation

```
java.lang.AssertionError: Expected 50 total invocations, but was 43 expected:<50> but was:<43>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at com.hazelcast.test.HazelcastTestSupport.assertEqualsStringFormat(HazelcastTestSupport.java:672)
	at com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorAbstractTest.assertTotalInvocations(SlowOperationDetectorAbstractTest.java:117)
	at com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest.testSlowRecursiveOperation(SlowOperationDetectorBasicTest.java:203)
```
https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OpenJDK7-Quality-Outreach/com.hazelcast$hazelcast/41/testReport/junit/com.hazelcast.spi.impl.operationexecutor.slowoperationdetector/SlowOperationDetectorBasicTest/testSlowRecursiveOperation/

```
java.lang.AssertionError: Expected 40 total invocations, but was 13. Log: {"operation":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation","stackTrace":"java.lang.Thread.sleep(Native Method)\n\tjava.lang.Thread.sleep(Thread.java:340)\n\tjava.util.concurrent.TimeUnit.sleep(TimeUnit.java:386)\n\tcom.hazelcast.test.HazelcastTestSupport.sleepSeconds(HazelcastTestSupport.java:288)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:309)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.recursiveCall(SlowOperationDetectorBasicTest.java:313)\n\tcom.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation.run(SlowOperationDetectorBasicTest.java:304)\n\tcom.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:183)\n\tcom.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:122)\n\tcom.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:102)","totalInvocations":13,"invocations":[{"id":1131959801,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=1131959801, partitionId=18, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933985962,"durationMs":3733},{"id":24405948,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=24405948, partitionId=27, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933985962,"durationMs":3733},{"id":389020642,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=389020642, partitionId=36, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933983899,"durationMs":5795},{"id":1985597259,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=1985597259, partitionId=9, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933985962,"durationMs":3733},{"id":515683530,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=515683530, partitionId=1, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933982894,"durationMs":3067},{"id":276515181,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=276515181, partitionId=10, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933982894,"durationMs":1004},{"id":1741639577,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=1741639577, partitionId=63, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933983899,"durationMs":5795},{"id":48760030,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=48760030, partitionId=81, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933983899,"durationMs":5795},{"id":1389100850,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=1389100850, partitionId=45, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933983899,"durationMs":5795},{"id":1448600620,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=1448600620, partitionId=54, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933983899,"durationMs":5795},{"id":1811408427,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=1811408427, partitionId=0, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933985962,"durationMs":3733},{"id":340058042,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=340058042, partitionId=19, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933982894,"durationMs":1004},{"id":257382505,"details":"com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest$SlowRecursiveOperation{serviceName='null', identityHash=257382505, partitionId=72, replicaIndex=0, callId=0, invocationTime=-1 (1970-01-01 01:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807}","startedAt":1465933983899,"durationMs":5795}]} expected:<40> but was:<13>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at com.hazelcast.test.HazelcastTestSupport.assertEqualsStringFormat(HazelcastTestSupport.java:923)
	at com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorAbstractTest.assertTotalInvocations(SlowOperationDetectorAbstractTest.java:131)
	at com.hazelcast.spi.impl.operationexecutor.slowoperationdetector.SlowOperationDetectorBasicTest.testSlowRecursiveOperation(SlowOperationDetectorBasicTest.java:198)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-sonar/com.hazelcast$hazelcast/1093/testReport/junit/com.hazelcast.spi.impl.operationexecutor.slowoperationdetector/SlowOperationDetectorBasicTest/testSlowRecursiveOperation/

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationexecutor/slowoperationdetector/SlowOperationDetectorAbstractTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationexecutor/slowoperationdetector/SlowOperationDetectorAbstractTest.java
@@ -256,17 +256,17 @@ abstract class SlowOperationDetectorAbstractTest extends HazelcastTestSupport {
         }
     }
 
-    static abstract class CountDownLatchOperation extends AbstractOperation {
+    static abstract class JoinableOperation extends AbstractOperation {
 
-        private final CountDownLatch latch = new CountDownLatch(1);
+        private final CountDownLatch completedLatch = new CountDownLatch(1);
 
         void done() {
-            latch.countDown();
+            completedLatch.countDown();
         }
 
-        void await() {
+        void join() {
             try {
-                latch.await();
+                completedLatch.await();
             } catch (InterruptedException e) {
                 EmptyStatement.ignore(e);
             }

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationexecutor/slowoperationdetector/SlowOperationDetectorBasicTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationexecutor/slowoperationdetector/SlowOperationDetectorBasicTest.java
@@ -26,6 +26,7 @@ import com.hazelcast.spi.properties.GroupProperty;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.test.annotation.Repeat;
 import org.junit.After;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
@@ -35,6 +36,9 @@ import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
 
+import static com.hazelcast.spi.properties.GroupProperty.PARTITION_OPERATION_THREAD_COUNT;
+import static com.hazelcast.spi.properties.GroupProperty.SLOW_OPERATION_DETECTOR_LOG_RETENTION_SECONDS;
+import static com.hazelcast.spi.properties.GroupProperty.SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS;
 import static java.lang.String.valueOf;
 
 @RunWith(HazelcastParallelClassRunner.class)
@@ -53,7 +57,7 @@ public class SlowOperationDetectorBasicTest extends SlowOperationDetectorAbstrac
     public void testDisabled() {
         Config config = new Config();
         config.setProperty(GroupProperty.SLOW_OPERATION_DETECTOR_ENABLED.getName(), "false");
-        config.setProperty(GroupProperty.SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS.getName(), "1000");
+        config.setProperty(SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS.getName(), "1000");
 
         instance = createHazelcastInstance(config);
 
@@ -102,7 +106,7 @@ public class SlowOperationDetectorBasicTest extends SlowOperationDetectorAbstrac
 
         SlowOperation operation = new SlowOperation(3);
         executeOperation(instance, operation);
-        operation.await();
+        operation.join();
 
         Collection<SlowOperationLog> logs = getSlowOperationLogsAndAssertNumberOfSlowOperationLogs(instance, 1);
 
@@ -118,7 +122,7 @@ public class SlowOperationDetectorBasicTest extends SlowOperationDetectorAbstrac
 
         SlowOperation operation = new SlowOperation(4, 2);
         executeOperation(instance, operation);
-        operation.await();
+        operation.join();
 
         Collection<SlowOperationLog> logs = getSlowOperationLogsAndAssertNumberOfSlowOperationLogs(instance, 1);
 
@@ -170,26 +174,26 @@ public class SlowOperationDetectorBasicTest extends SlowOperationDetectorAbstrac
         int numberOfOperations = 40;
         int recursionDepth = 15;
 
-        Config config = new Config();
-        config.setProperty(GroupProperty.SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS.getName(), "1000");
-        config.setProperty(GroupProperty.SLOW_OPERATION_DETECTOR_LOG_RETENTION_SECONDS.getName(), valueOf(Integer.MAX_VALUE));
-        config.setProperty(GroupProperty.PARTITION_OPERATION_THREAD_COUNT.getName(), valueOf(partitionThreads));
+        Config config = new Config()
+                .setProperty(SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS.getName(), "1000")
+                .setProperty(SLOW_OPERATION_DETECTOR_LOG_RETENTION_SECONDS.getName(), valueOf(Integer.MAX_VALUE))
+                .setProperty(PARTITION_OPERATION_THREAD_COUNT.getName(), valueOf(partitionThreads));
         instance = createHazelcastInstance(config);
 
         List<SlowRecursiveOperation> operations = new ArrayList<SlowRecursiveOperation>(numberOfOperations);
-        int partitionCount = getNode(instance).nodeEngine.getPartitionService().getPartitionCount();
+        int partitionCount = getPartitionService(instance).getPartitionCount();
 
         int partitionIndex = 1;
         for (int i = 0; i < numberOfOperations; i++) {
             int partitionId = partitionIndex % partitionCount;
-            SlowRecursiveOperation operation = new SlowRecursiveOperation(partitionId, recursionDepth, 4);
+            SlowRecursiveOperation operation = new SlowRecursiveOperation(partitionId, recursionDepth, 20);
             operations.add(operation);
             executeOperation(instance, operation);
             partitionIndex += partitionCount / partitionThreads + 1;
         }
 
         for (SlowRecursiveOperation operation : operations) {
-            operation.await();
+            operation.join();
         }
 
         Collection<SlowOperationLog> logs = getSlowOperationLogsAndAssertNumberOfSlowOperationLogs(instance, 1);
@@ -221,7 +225,7 @@ public class SlowOperationDetectorBasicTest extends SlowOperationDetectorAbstrac
         }
     }
 
-    private static class SlowOperation extends CountDownLatchOperation {
+    private static class SlowOperation extends JoinableOperation {
 
         private final int sleepSeconds;
 
@@ -283,11 +287,11 @@ public class SlowOperationDetectorBasicTest extends SlowOperationDetectorAbstrac
         }
 
         public void await() {
-            operation.await();
+            operation.join();
         }
     }
 
-    private static class SlowRecursiveOperation extends CountDownLatchOperation {
+    private static class SlowRecursiveOperation extends JoinableOperation {
 
         private final int recursionDepth;
         private final int sleepSeconds;

